toolchain/lower/function_context.h

```diff
@@ -203,6 +225,13 @@ class FunctionContext {
   // The optional vlog stream.
   llvm::raw_ostream* vlog_stream_;
 
+  // This is initialized and populated while lowering a specific function.
+  // A "state" corresponds to an inst_id. If anything specific-dependent
+  // was found, the state is added to the list of states: current_states_.
+  // The storage for current_states_ is in the file_context.
+  FileContext::LoweringFunctionState current_state_;
+  llvm::SmallVector<FileContext::LoweringFunctionState>* current_states_;
```

Comment:
It's subtle, but can `current_states_` be null when not lowering a specific?

Also, other thought while reading this, might consider using `current_states_.back()` instead of a local here -- not sure how much you gain from the separation, particularly given the trade-off of 3 array relocations